include/CL/sycl/buffer/detail/buffer.hpp

```diff
@@ -25,6 +25,20 @@ namespace cl {
 namespace sycl {
 namespace detail {
 
+/** Used to retrieve the type return by an Iterator,
+ *  in particular a const-iterator<T> returns const T.
+ *  The attribute is_const is true iff the iterator is
+ *  a const one
+ */
+template<typename Iterator>
+struct iterator_value_type
+{
+  using value_type  = typename std::remove_pointer_t<
+    typename std::iterator_traits<Iterator>::pointer>;
+  constexpr static bool is_const = std::is_const<value_type>::value;
```

Comment:
Could be also just `std::is_const<std::iterator_traits<Iterator>::value_type>::value` with some `lacking typename`.
But after some thought, I have the feeling that it does not work for some output iterators when there is no value type at all http://en.cppreference.com/w/cpp/concept/OutputIterator